### PR TITLE
feat: add webdriver bidi support

### PR DESCRIPTION
**Pull request by jlipps**
This brings bidi support to the chromium driver. It depends on:

* https://github.com/appium/appium-chromedriver/pull/362
* https://github.com/appium/appium/pull/19404

### CODE REVIEW

File: lib/driver.js [hunk 1]
```diff
@@ -11,7 +11,8 @@ const CD_CONSTRAINTS = /** @type {const} */ ({
   logPath: { isString: true },
   autodownloadEnabled: { isBoolean: true },
   disableBuildCheck: { isBoolean: true },
-  browserName: { isString: true }
+  browserName: { isString: true },
+  webSocketUrl: { isBoolean: true },
```
Comment (mykola-mokhnach) [inline lib/driver.js:15]: should we add this property to the base driver as it is defined by the standard?

File: lib/driver.js [hunk 2]
```diff
@@ -46,6 +52,11 @@ export class ChromiumDriver extends BaseDriver {
     return true;
   }
 
+  getBidiProxy() {
```
Comment (mykola-mokhnach) [inline lib/driver.js:55]: should this be a simple getter?

File: lib/driver.js [hunk 3]
```diff
@@ -59,14 +70,19 @@ export class ChromiumDriver extends BaseDriver {
     jsonwpRequiredCaps,
     w3cCapabilities,
   ) {
-    const [sessionId, caps] = /** @type {[string, ChromiumDriverCaps]} */(await super.createSession(jsonwpDesiredCapabilities, jsonwpRequiredCaps, w3cCapabilities));
-    await this.startChromedriverSession();
-    return [sessionId, caps];
+    const [sessionId,] = /** @type {[string, ChromiumDriverCaps]} */(await super.createSession(jsonwpDesiredCapabilities, jsonwpRequiredCaps, w3cCapabilities));
+    const returnedCaps = await this.startChromedriverSession();
+    if (returnedCaps.webSocketUrl) {
+      // @ts-ignore while webSocketUrl is a boolean when sent by the client, it's a string when it
```
Comment (mykola-mokhnach) [inline lib/driver.js:76]: can we just cast it explicitly to string?

File: lib/driver.js [hunk 4]
```diff
@@ -88,10 +104,11 @@ export class ChromiumDriver extends BaseDriver {
     };
 
     this.cd = new Chromedriver(cdOpts);
-    await this.cd.start(this.getSessionCaps());
+    const cdStartRes = /** @type {ChromiumDriverCaps} */(await this.cd.start(this.getSessionCaps()));
```
Comment (mykola-mokhnach) [inline lib/driver.js:107]: What type cd.start returns? linter complains about it is currently void

Comment (jlipps) [reply lib/driver.js:107]: this was due to the chromedriver dep. now that the linked PR is merged and published, this is not an issue.

File: lib/driver.js (hunk 1 above)
Comment (jlipps) [reply lib/driver.js:15]: it's already there!

File: lib/driver.js (hunk 2 above)
Comment (jlipps) [reply lib/driver.js:55]: sure. i didn't think TS could handle this well but it seems it can.

File: lib/driver.js (hunk 3 above)
Comment (jlipps) [reply lib/driver.js:76]: you have to do this:

```js
this._bidiProxyUrl = (/** @type {string|null} */(/** @type {unknown} */(returnedCaps.webSocketUrl)));
```

not sure it's better 🤷

File: lib/driver.js [hunk 5]
```diff
@@ -59,14 +69,17 @@ export class ChromiumDriver extends BaseDriver {
...
+    if (returnedCaps.webSocketUrl) {
+      this._bidiProxyUrl = (/** @type {string|null} */(/** @type {unknown} */(returnedCaps.webSocketUrl)));
+    }
+    return [sessionId, returnedCaps];
   }
 
   getSessionCaps() {
     return Object.keys(this.opts).reduce((acc, capName) => {
-      if (STANDARD_CAPS_LOWER.has(capName) || capName.startsWith(CHROME_VENDOR_PREFIX) || capName.startsWith(EDGE_VENDOR_PREFIX)) {
+      if (STANDARD_CAPS_LOWER.has(capName.toLowerCase()) || capName.startsWith(CHROME_VENDOR_PREFIX) || capName.startsWith(EDGE_VENDOR_PREFIX)) {
```
Comment (mykola-mokhnach) [inline lib/driver.js:82]: cap names are case sensitive, aren't they?

File: lib/driver.js [hunk 6]
```diff
@@ -59,14 +69,17 @@ export class ChromiumDriver extends BaseDriver {
     jsonwpRequiredCaps,
     w3cCapabilities,
   ) {
-    const [sessionId, caps] = /** @type {[string, ChromiumDriverCaps]} */(await super.createSession(jsonwpDesiredCapabilities, jsonwpRequiredCaps, w3cCapabilities));
-    await this.startChromedriverSession();
-    return [sessionId, caps];
+    const [sessionId,] = /** @type {[string, ChromiumDriverCaps]} */(await super.createSession(jsonwpDesiredCapabilities, jsonwpRequiredCaps, w3cCapabilities));
```
Comment (mykola-mokhnach) [inline lib/driver.js:72]: actually all 3 arg names there are `w3cCapabilities`

**Review by mykola-mokhnach — APPROVED** [review]

File: lib/driver.js (hunk 6 above)
Comment (jlipps) [reply lib/driver.js:72]: ah good point. can fix later.

File: lib/driver.js (hunk 5 above)
Comment (jlipps) [reply lib/driver.js:82]: mm, I needed this to fix an issue where caps were not properly forwarded to chromedriver.

**Pull request merged**